src/IssueViz/Selection/IssueSelectionService.cs

```diff
@@ -0,0 +1,87 @@
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Selection
+{
+    /// <summary>
+    /// Represents the currently selected <see cref="IAnalysisIssueVisualization"/>.
+    /// Raises an event <see cref="SelectedIssueChanged"/> when selection changes.
+    /// </summary>
+    public interface IIssueSelectionService : IDisposable
+    {
+        event EventHandler SelectedIssueChanged;
+
+        IAnalysisIssueVisualization SelectedIssue { get; set; }
+    }
+
+    [Export(typeof(IIssueSelectionService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class IssueSelectionService : IIssueSelectionService
+    {
+        private readonly IAnalysisIssueSelectionService flowStepSelectionService;
+        private IAnalysisIssueVisualization selectedIssue;
+
+        [ImportingConstructor]
+        public IssueSelectionService(IAnalysisIssueSelectionService flowStepSelectionService)
+        {
+            this.flowStepSelectionService = flowStepSelectionService;
+
+            flowStepSelectionService.SelectionChanged += FlowStepSelectionService_SelectionChanged;
+        }
+
+        public event EventHandler SelectedIssueChanged;
+
+        public IAnalysisIssueVisualization SelectedIssue
+        {
+            get => selectedIssue;
+            set
+            {
+                var selectionChanged = selectedIssue != value;
+
+                if (selectionChanged)
+                {
+                    selectedIssue = value;
+                    SelectedIssueChanged?.Invoke(this, EventArgs.Empty);
+
+                    if (flowStepSelectionService.SelectedIssue != value)
+                    {
+                        var hasSecondaryLocations = value != null && value.Flows.SelectMany(x => x.Locations).Any();
+                        flowStepSelectionService.SelectedIssue = hasSecondaryLocations ? value : null;
+                    }
+                }
+            }
+        }
+
+        private void FlowStepSelectionService_SelectionChanged(object sender, SelectionChangedEventArgs e)
```

Comment:
We might not need to subscribe to this event after everything is done.
We can update the IssueViz using the old service, so that's why I'm syncing the changes, but in the future I hope that this would be the only service that can set the IssueViz, and the detailed one could only read it, and set the flow/step. When that would be the case, we won't need to listen to this event.

Reply:
We shouldn't need to - if we do, then there's probably something wrong with the design.
I think it's worth putting a design comment in the code saying this should become redundant.

If I remember correctly, we thought the flow panel only sets the selection to null if the primary span becomes null, which is a hack that should disappear.

It looks like a fair chunk of the tests are about ensure this hacky behaviour still works for now.
I think it would have been ok not to handle this case here, even if it means breaking that particularly hacky scenario in the short term.

Reply:
Yep I expect all these tests to go away as well. I was debating if I should just skip them, but in the end decided to keep them for now, so we would have an option to delay the fixing the hack if we'd need to.